docs/component-stability.md

```diff
@@ -66,6 +66,99 @@ Stable components MUST be compatible between minor versions unless critical secu
+constraints are set out in this document.
+
+For pipeline components however, this section details the kinds of values that should be observable
+via internal telemetry for all stable components.
+
+**Notes:**
+- The following categories MUST all be covered, unless justification is given as to why
+one may not be applicable.
+
+- However, for each category, many reasonable implementations are possible, as long as the relevant
+information can be derived from the emitted telemetry; everything after the basic category
+description is a recommendation, and is not normative.
+
+- Of course, a component may define additional internal telemetry which is not in this list.
+
+- Some of this internal telemetry may already be provided by pipeline auto-instrumentation or
+helper modules (such as `receiverhelper`, `scraperhelper`, `processorhelper`, or `exporterhelper`).
+Please check the documentation to verify which parts, if any, need to be implemented manually.
+
+**Definition:** In the following, an "item" refers generically to a single log record, metric point,
+or span.
+
+The internal telemetry of a stable pipeline component should allow observing the following:
+
+1. How much data the component receives.
+
+    For receivers, this could be a metric counting requests, received bytes, scraping attempts, etc.
+
+    For other components, this would typically be the number of items received through the
+    `Consumer` API.
+
+2. How much data the component outputs.
+
+    For exporters, this could be a metric counting requests, sent bytes, etc.
+
+    For other components, this would typically be the number of items forwarded to the next
+    component through the `Consumer` API.
+
+3. How much data is dropped because of errors.
+
+    For receivers, this could include a metric counting payloads that could not be parsed in.
+
+    For receivers and exporters that interact with an external service, this could include a metric
+    counting requests that failed because of network errors.
+
+    The goal is to be able to easily pinpoint the source of data loss in the Collector pipeline, so
+    this should either:
+    - only include errors internal to the component, or;
+    - allow distinguishing said errors from ones originating in an external service, or propagated
+        from downstream Collector components.
+
+4. Details for error conditions.
+
+    This could be in the form of logs or spans detailing the reason for an error. As much detail as
+    necessary should be provided to ease debugging. Processed signal data should not be included for
+    security and privacy reasons.
+
+5. Other possible discrepancies between input and output, if any. This may include:
+
+    - How much data is dropped as part of normal operation (eg. filtered out).
+
+    - How much data is created by the component.
+
+    - How much data is currently held by the component (eg. an UpDownCounter keeping track of the
+        size of an internal queue).
```

Comment:
is it worth defining a set of internal telemetry for queueing / batching? i.e. things that hold on to items?

Reply:
Do you mean we should define *more specific* requirements? It sounds like many components may have an internal queue, so I don't want to increase the component author's burden too much. But we could perhaps suggest a gauge for the queue's capacity on top of its current size?

Or do you mean that maybe this category is not worth it? For the record, we already emit similar telemetry for the internal queues of exporters relying on `exporterhelper`.

Reply:
yeah more specific requirements, essentially codifying how exporterhelper does this today to make it clear to end-users where component authors are using internal queues. This is really vital for understanding backpressure from an end-user perspective IMO.

Reply:
I added a requirement to emit the queue's capacity (if it is fixed), and added "size as an updowncounter and capacity as a gauge" (the current `exporterhelper` implementation) as the example implementation.